### DISCUSSION TITLE
`TypeIs` and unions of `Callable`

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#9282] [Q&A]
Code sample in [pyright playground](https://pyright-play.net/?pyrightVersion=1.1.384&strict=true&enableExperimentalFeatures=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANFAMICGANtZQEbUCmZAKokwJIDOAUPwBMmwTNwD63VGmZiAxjTqMmAClkAuCgobMA2joC6ZInH1QAPptrame1DEOFi%2BgJRQAtAD4o7BF246qKyVbFHsjJ301XigYqAA6BP5eIREYJm4YVQ1AxV0DcJNzS1ybHTsHYxco2MwRJHFJdBl5IOZVZ2qa2JAmADcmGjF4X3bo2KZqbiZOrqge-sHhlVlnIA)

```python
from typing import Any, Callable, TypeIs


def is_single_callable(c: Callable[[], Any] | Callable[[int], Any]) -> TypeIs[Callable[[int], Any]]:
    ...


def test(c: Callable[[], Any] | Callable[[int], Any]):
    if is_single_callable(c):
        reveal_type(c)  # (int) -> Any
    else:
        reveal_type(c)  # (() -> Any) | ((int) -> Any)
```

I'm wondering if Pyright can safely narrow `c` to `((int) -> Any)` in the `else` branch

**Comment by erictraut** [comment]
I think it would be type safe to narrow it further in the fall-through case. The question is whether such an enhancement is worthwhile. Pyright's current behavior is arguably correct, but the resulting type could be more precise (narrower). Mypy appears to over-narrow the type to `Never` in this case, which I would consider a bug.

**Reply by Viicos** [reply]
Yes I also noticed this in mypy, jus opened an issue.

> The question is whether such an enhancement is worthwhile.

Honestly I can leave without it, implementing it might be a bit painful as callable signatures may overlap in some way. I mainly wanted to confirm this wasn't by design and could be safely narrowed. Thanks!
